### DISCUSSION TITLE
workmode Peak Shaving

### DISCUSSION THREAD

**Discussion by chrigu62** [nathanmarlor/foxess_modbus#699] [Support]
I can't select workmode Peak Shaving and the integration shows noting in the list when Peak Shaving is activated in App.
I have inverter H3-10.0-E master firmware 2.01

**Comment by KNZZZZZZZ** [comment]
Same for me on the H1-5.0-E-G2, any updates if Peak Shaving mode will be available on this integration?
Thanks

**Comment by miralemR** [comment]
Also with my H3. From 1. Okt it's crucial function to reduce energy costs. Shall we see any changes in the future?

**Comment by FozzieUK** [comment]
Would one of you please put your H3 inverter into peak shaving mode, and the do a read registers on address 41000 and post back the response please.

i.e. in HA Developer Tools, Actions - select Foxess Modbus - Read Registers set register address 41000, count 1, holding

@KNZZZZZZZ I wasn't aware the H1 G2 had a peak shaving function ?

**Reply by KNZZZZZZZ** [reply]
> @KNZZZZZZZ I wasn't aware the H1 G2 had a peak shaving function ?

Yes it's also available on the H1 but only on the G2 version.
Should I also test the same register address?

**Reply by FozzieUK** [reply]
yes please

**Reply by KNZZZZZZZ** [reply]
I also got the value 4 for 41000

**Reply by FozzieUK** [reply]
Ok, thanks if you are on v1.12.2 you can try the change I listed a few posts below if you are on the latest v1.13.0 the line number will be 2101 - that should work for you. 

If you can confirm that it does work I will raise the PR to add it properly.

**Reply by KNZZZZZZZ** [reply]
I did the modification and the working mode change works perfectly (HA Modbus v1.12.2).

I tested the other values ​​too on my H1-G2 (Manager 1.48, Master 1.30, Slave 1.03):

values:
  "41001": 1
  "41002": 0
  "41003": 0
  "41004": 1
  "41005": 0
  "41006": 0


values:
  "46501": 0
  "46502": 2400     --> That is indeed the import limit I had set

values:
  "46503": 20     --> That is indeed the threshold SoC I had set

values:
  "46504": 0
  "46505": 0

values:
  "46506": 0

values:
  "46507": 0

values:
  "46508": 0

values:
  "46509": 0

values:
  "46510": 0

Not sure what ChrInLowImport mean as I don't use it.
Export Peak Limit seems interesting to me with a dynamic rate if you want to limit export when it costs you money to export instead of a fee?

**Reply by FozzieUK** [reply]
Ok thanks for confirming, the G2 and H3 Pro seem to have very closely aligned register maps.

**Reply by KNZZZZZZZ** [reply]
Will the import limit, threshold soc and export peak limit setting also be added or how can I do this myself? Thanks!

**Reply by FozzieUK** [reply]
Not at this moment, that will need a much bigger piece of work to test and develop.
but as you know the addresses you should be able to create an automation that will read them using the action Foxess Modbus - read registers and write to them using write registers.

This for example is reading the real time clock and taking the returned values out of the response

![11](https://github.com/user-attachments/assets/3b09e9e3-c836-4b3c-976b-9c73d194132c)

**Comment by miralemR** [comment]
Hello,

I put my H3 inverter into peak shaving mode, read register 41000, and got the value 4. Let me know if this matches the expected behavior or if further testing is needed.

**Reply by FozzieUK** [reply]
Great, thank you - can you let me know what version of the integration you are running, and I can give a small code change for you to test the peak-shaving work mode.

**Reply by miralemR** [reply]
I hope that I got it right... I’m running version v1.12.2 of the integration.

**Reply by FozzieUK** [reply]
Also sorry to be a pain but if you could repeat that register read on 41000 for each work mode you have available on the website drop down - i'm seeing Self Use (0), Feed In (1) and Backup (2) so it seems there may be a missing mode (3) ?

**Reply by miralemR** [reply]
No problem... glad to help. Just need a few minutes to check

**Reply by miralemR** [reply]
I had issues while switching the work mode... I got disconnected from the grid because I’m not allowed to export electricity to the grid.

In self-use mode, it’s 0,
Feed-in first is 1,
Backup is 2.
The drop-down menu also offers force charge and force discharge.
The outage happened when I tried force discharge mode, so it took me some time to respond

**Reply by miralemR** [reply]
In the menu directly on the inverter, there are only the three options you mentioned and peak shaving. Force discharge and force charge are available in the drop-down menu in only in Foxess Modbus

**Reply by FozzieUK** [reply]
Ok thanks very much for that (sorry about your disconnection - hope that doesn't cause you too much trouble).

If you wouldn't mind trying this when you have a spare 30 minutes, by modifying your local HA installation and change the file 
`custom_components/foxess_modbus/entities/entity_descriptions.py` as follows 

If you are on V1.12.2 then on Line 2015 you will find the H3_SET work mode def:-

```
options_map={0: "Self Use", 1: "Feed-in First", 2: "Back-up"},
```

Can you change it to this -

```
options_map={0: "Self Use", 1: "Feed-in First", 2: "Back-up", 4: "Peak Shaving"},
```

i.e. you are adding the text `, 4: "Peak Shaving"`

Save the file and restart your home assistant.

If you could then check what options you have in the workmode select, and try switching to Peak Shaving mode and check on the inverter that it has changed, then switch back to Self Use and again check it has changed correctly.

**Reply by miralemR** [reply]
I followed your instructions, and it works as you expected it would

**Reply by FozzieUK** [reply]
Great, thanks for your help 👍  - i'll wait to see what the H1 G2 also does and then i'll do a PR to close this off.

**Reply by miralemR** [reply]
Great... and thank you.

I would like to ask one more thing… Is it possible to enable access through Foxess Modbus in Home Assistant to the following settings:

Settings/Workmode/Charge Time
Settings/Workmode/PeakShavingSet (Import limit, Threshold SOC, Export peak limit, ChrInLowImport/(NO Mode, NC Mode, Timing Mode), ChrInLowTime)?

**Reply by FozzieUK** [reply]
The charge times were never available on the H3, I don't think they will be added as Fox appear to be moving away from them to the scheduler but you're welcome to test to see.

They have to be all read together so if you try doing a register read starting at 41001, count 6, holding and see if that returns anything.

On the Peak Shaving, these registers were not available on the H3, but they are included within the latest Fox modbus spec (typically used by the H3 Pro and H1 G2) - it is possible they have been added to the H3 in latest firmware.

You could try doing these test register reads and see what response you get -

Import Limit = 46501, count=2, holding
Threshold SoC = 46503, count=1, holding
Export Peak Limit = 46504, count=2, holding
ChrInLowImport = 46506, count=1, holding
ChrInLowTime1StartHour = 46507, count=1, holding
ChrInLowTime1StartMinute = 46508, count=1, holding
ChrInLowTime1EndHour = 46509, count=1, holding
ChrInLowTime1StartMinute = 46510, count=1, holding

I can't find any register definitions for NO/NC or Timing Mode.

**Reply by miralemR** [reply]
For first 
values:
  "41001": 0
  "41002": 0
  "41003": 0
  "41004": 0
  "41005": 0
  "41006": 0

For Treshold soc and every other 

error: >-
  Error reading registers. Type: RegisterType.HOLDING; start: 46503; count: 1;
  slave: 247 from tcp://192.168.68.78:502: Exception Response(131, 3,
  IllegalAddress)
Does that mean it's not possible to set Treshold soc through Home Assistant?

**Reply by FozzieUK** [reply]
Yes unfortunately it does look as if you can’t set the threshold soc via modbus.

On the charge times, do you have any charge periods set? - if not could you set one for any time (it doesn’t have to be on now just set) and then try reading the 41001 registers again please.

**Reply by miralemR** [reply]
I set timers and enabled them... same as before 
values:
  "41001": 0
  "41002": 0
  "41003": 0
  "41004": 0
  "41005": 0
  "41006": 0
Just say if need anything else

**Reply by FozzieUK** [reply]
Ok, thanks for confirming.
It was worth trying but unfortunately the charge periods are still not supported on the H3 modbus as you should see 41001 go to 1 when period 1 is enabled, 41002 is the encoded start time, 41003 is the encoded end time (and likewise for period 2 in 41004-41006)
Obviously force charge work mode is available but you’ll have to manage the start and end times in HA.

**Comment by miralemR** [comment]
Actually I don't need charge set periods under 
Work mode/Charge time. Nor I switch between self-use and peak shaving... always on peak shaving. 

On inverter... only when I select Settings/Workmode/PeakShaving 
it shows me the option 
Settings/Workmode/PeakShavingSet
and under that I can set Import limit, Threshold SOC, Export peak limit, ChrInLowImport and ChrInLowTime. For my automation in HA I need just Import limit and Threshold SOC to manipulate. 
If I understand correctly that this cannot be achieved with Foxess Modbus, then I don't know how else I could do it in HA. Any suggestions? 
One more thing... otherwise I have access to these settings through my Android app from my supplier. Just can't make it setting automatically according to Solcast

**Reply by FozzieUK** [reply]
I’ve looked through all of the modbus protocol documents and other than on the latest spec (for the H3 Pro and G2) there are no modbus registers on the H3 to set them.

The H3 inverter is hugely popular and it is more likely that Fox may add these registers in a future firmware - the only suggestion is to raise a ticket on Fox service and ask them if they will add the Peak Shaving registers on modbus as they may already have this on their dev list.
